common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -73,45 +72,50 @@ private static long toLong(String s) {
    * This method is case-insensitive.
    */
   public static CalendarInterval fromString(String s) {
-    if (s == null) {
-      return null;
-    }
-    s = s.trim();
-    Matcher m = p.matcher(s);
-    if (!m.matches() || s.compareToIgnoreCase("interval") == 0) {
+    try {
+      return fromCaseInsensitiveString(s);
+    } catch (IllegalArgumentException e) {
       return null;
-    } else {
-      long months = toLong(m.group(1)) * 12 + toLong(m.group(2));
-      long microseconds = toLong(m.group(3)) * MICROS_PER_WEEK;
-      microseconds += toLong(m.group(4)) * MICROS_PER_DAY;
-      microseconds += toLong(m.group(5)) * MICROS_PER_HOUR;
-      microseconds += toLong(m.group(6)) * MICROS_PER_MINUTE;
-      microseconds += toLong(m.group(7)) * MICROS_PER_SECOND;
-      microseconds += toLong(m.group(8)) * MICROS_PER_MILLI;
-      microseconds += toLong(m.group(9));
-      return new CalendarInterval((int) months, microseconds);
     }
   }
 
   /**
-   * Convert a string to CalendarInterval. Unlike fromString, this method can handle
+   * Convert a string to CalendarInterval. This method can handle
    * strings without the `interval` prefix and throws IllegalArgumentException
    * when the input string is not a valid interval.
    *
    * @throws IllegalArgumentException if the string is not a valid internal.
    */
   public static CalendarInterval fromCaseInsensitiveString(String s) {
```

Comment:
Is it the only place we parse interval string? I thought we parse it with antlr parser.

Reply:
antlr parser does this as well but it parses sql elements like
```sql
spark-sql> select interval 10 days 1 second;
interval 1 weeks 3 days 1 seconds
```
here is only the place where we parse string values:
```sql
spark-sql> select interval 'interval 10 days 1 second';
interval 1 weeks 3 days 1 seconds
```

Reply:
This looks duplicated. Shall we add a `parseInterval` method to the `ParserInterface` interface and call the parser here?

Reply:
Maybe something has been duplicated, and can be reused but this is heavy refactoring for this PR. 

For instance, `AstBuilder.visitInterval` gets already split interval units but `CalendarInterval.fromString()` uses regular expression to parse & split: https://github.com/apache/spark/blob/b10344956d672fca495d0684b351ab5cf9f210d8/common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java#L50-L53

If you don't mind, I would try to do that in a separate PR.

Reply:
And your regexp is not tolerant to the order of interval units, see:
```sql
spark-sql> select interval 'interval 1 microsecond 2 months';
NULL
spark-sql> select interval 1 microsecond 2 months;
interval 2 months 1 microseconds
```

Reply:
Let's keep them separate so far. And I will try to write flexible and common code for parsing string interval that could handle other features found in https://github.com/apache/spark/pull/26055

Reply:
SGTM